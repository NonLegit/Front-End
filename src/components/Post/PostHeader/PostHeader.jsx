// mui components
import {
  Avatar, Box, Typography,
} from '@mui/material';
import CallSplitIcon from '@mui/icons-material/CallSplit';
import calculateTime from '../../../utils/calculateTime';

// styles
import {
  CreatedAt,
  Flair,
  PostInfo, PostInfoLink, PostTitle, Tag,
} from './styles';
/**
 * This component is the upper section of post
 *
 * @component PostHeader
 * @property {string} title -Post title.
 * @property {string} ownerType -Post owner type user or subreddit.
 * @property {string} ownerIcon -Post ownerName icon.
 * @property {string} ownerName -Post subreddit(post ownerName).
 * @property {string} authorName -Post authorName.
 * @property {string} flairText -Post flair text.
 * @property {string} flairBackgroundColor -Post flair background color.
 * @property {string} flairColor -Post flair color.
 * @property {boolean} subredit -to identify if post in home page or subreddit.
 * @property {boolean} nsfw -Whether the post is not safe for work or not.
 * @property {boolean} spoiler -Whether the post is spoiler or not.
 * @returns {React.Component} Post header
 */

function PostHeader(props) {
  const {
    title, ownerIcon, ownerType, ownerName, authorName, flairText, flairBackgroundColor, flairColor, createdAt,
    subredit, nsfw, spoiler, redirectToPost, sharedFrom,
  } = props;

  const handleClickOnTitle = (e) => {
    e.preventDefault();
    redirectToPost(true, false);
  };

  return (
    <>
      <PostInfo pb={1}>
        {!subredit
        && (
          <>
            <Avatar
              src={ownerIcon}
              sx={{
                width: 20,
                height: 20,
              }}
              alt="Profile ownerIcon"
            />
            <PostInfoLink to={ownerType === 'Subreddit' ? `/r/${ownerName}` : `/user/${ownerName}`} color="#000" fontWeight="bolder">
              {ownerType === 'Subreddit' ? 'r/' : 'u/'}
              {ownerName}
            </PostInfoLink>
          </>
        )}
        <Box color="#787C7E" fontWeight={300} display="flex" gap="4px" flexWrap="wrap">
          {ownerType === 'Subreddit' && (
          <>
            {!subredit && (
            <span>
              â€¢
            </span>
            )}
            {sharedFrom
              ? (
                <>
                  <CallSplitIcon
                    sx={{
                      color: '#0079D3',
                      transform: 'rotate(90deg)',
                      width: 15,
                      height: 15,
                    }}
                  />
                  <div>Crossed By</div>
                </>
              )
              : <div>Posted By</div>}
            <PostInfoLink to={`/user/${authorName}`} color="inherit" fontWeight="normal">
              u/
              {authorName}
            </PostInfoLink>
          </>
          )}

          <CreatedAt color="inherit" fontWeight="normal">
            {calculateTime(createdAt)}
          </CreatedAt>
        </Box>
      </PostInfo>
      <PostTitle onClick={handleClickOnTitle}>
        <Typography
          variant="h6"
          component="h3"
          sx={{ fontSize: 18 }}
        >
          {title}
          {' '}
          {
            flairText
            && (
            <Flair
              disableRipple
              backgroundColor={flairBackgroundColor}
              flairColor={flairColor}
            >
              {flairText}
            </Flair>
            )
          }
          {' '}
          {
            spoiler && (
            <Tag
              color="third"
              variant="outlined"
            >
              spoiler
            </Tag>
            )
          }
          {
            nsfw && (
            <Tag
              color="nsfw"
              variant="outlined"
            >
              nsfw
            </Tag>
            )
          }
        </Typography>
      </PostTitle>
    </>
  );
}

export default PostHeader;
