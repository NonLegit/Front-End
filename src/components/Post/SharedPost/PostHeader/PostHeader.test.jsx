import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { MemoryRouter as Router } from 'react-router-dom';
import PostHeader from './PostHeader';
import calculateTime from '../../../../utils/calculateTime';

describe(PostHeader, () => {
  it('PostHeader renders correctly', () => {
    const testCase = {
      title: 'title', image: 'image', owner: 'owner', author: 'author', flair: 'flair', flairBackgroundColor: 'flairBackgroundColor', flairColor: 'flairColor',
    };
    const tree = renderer
      .create(
        <Router>
          <PostHeader
            title={testCase.title}
            image={testCase.image}
            owner={testCase.owner}
            author={testCase.author}
            flair={testCase.flairText}
            flairBackgroundColor={testCase.flairBackgroundColor}
            flairColor={testCase.flairColor}
          />
        </Router>
        ,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe(calculateTime, () => {
  const second = 1000;
  const minute = second * 60;
  const hour = minute * 60;
  const day = hour * 24;
  const week = day * 7;
  const month = day * 30;
  const year = day * 365;

  const today = new Date();
  const testCases = [{
    input: today - second,
    output: 'a few seconds ago',
  },
  {
    input: today - 2 * second,
    output: 'a few seconds ago',
  },
  {
    input: today - 60 * second,
    output: 'a minute ago',
  },
  {
    input: today - minute,
    output: 'a minute ago',
  },
  {
    input: today - 2 * minute,
    output: '2 minutes ago',
  },
  {
    input: today - 120 * minute,
    output: '2 hours ago',
  },
  {
    input: today - hour,
    output: 'an hour ago',
  },
  {
    input: today - 21 * hour,
    output: '21 hours ago',
  },
  {
    input: today - 23 * hour,
    output: 'a day ago',
  },
  {
    input: today - 2 * hour,
    output: '2 hours ago',
  },
  {
    input: today - 24 * hour,
    output: 'a day ago',
  },
  {
    input: today - 1 * day,
    output: 'a day ago',
  },
  {
    input: today - 2 * day,
    output: '2 days ago',
  },
  {
    input: today - 7 * day,
    output: '7 days ago',
  },
  {
    input: today - 14 * day,
    output: '14 days ago',
  },
  {
    input: today - week,
    output: '7 days ago',
  },
  {
    input: today - 2 * week,
    output: '14 days ago',
  },
  {
    input: today - 30 * day,
    output: 'a month ago',
  },
  {
    input: today - month,
    output: 'a month ago',
  },
  {
    input: today - 2 * month,
    output: '2 months ago',
  },
  {
    input: today - 365 * day,
    output: 'a year ago',
  },
  {
    input: today - year,
    output: 'a year ago',
  },
  {
    input: today - 2 * year,
    output: '2 years ago',
  },
  ];
  testCases.forEach((testCase, index) => {
    it(`Test Case #${index} : calculateTime works correctly with input = ${new Date(testCase.input).toISOString()}`, () => {
      expect(calculateTime(new Date(testCase.input).toISOString())).toBe(testCase.output);
    });
  });
});
