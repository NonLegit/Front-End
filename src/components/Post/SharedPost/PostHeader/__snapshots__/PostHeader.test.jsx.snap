// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostHeader PostHeader renders correctly 1`] = `
Array [
  <div
    className="MuiBox-root css-1eogohz"
  >
    <a
      className="css-kqg5s"
      color="#000"
      fontWeight="bolder"
      href="/user/undefined"
      onClick={[Function]}
    >
      u/
    </a>
    <div
      className="MuiBox-root css-1clzuz0"
    >
      <div
        className="MuiBox-root css-5ah3mn"
      >
        a few seconds ago
      </div>
    </div>
  </div>,
  <a
    className="css-10m9jmh"
    href="/"
    onClick={[Function]}
  >
    <h3
      className="MuiTypography-root MuiTypography-h6 css-6aghek-MuiTypography-root"
    >
      title
       
       
    </h3>
  </a>,
]
`;
